lib/http/request/_request_timing.js

```diff
@@ -0,0 +1,116 @@
+'use strict';
+
+var times = {};
+var start = function() {};
+var createResponseCallback = function() {};
+var createSocketCallback = function() {};
+var _ENV = process.env;
+
+var dogStatsDHost = _ENV.CN_DATADOG_HOST || '127.0.0.1';
+var dogStatsDPort = _ENV.CN_DATADOG_PORT || 8120;
+
+if (_ENV.HTTP_REQUEST_TIMING) {
+
+  var DogStatsD = require('hot-shots');
+  var statsClient = new DogStatsD({
+    host: dogStatsDHost,
+    port: dogStatsDPort,
+    maxBufferSize: 512
+  });
+
+  /**
+   * Take a time reading with a specific name, as a delta
+   * from another mark. The second argment is an array of mark names,
+   * so you can specify the time taken to 'firstByte' as as delta
+   * from 'tlsHandshake' for HTTPS connections, or from 'tcpConnection'
+   * for HTTP connections.
+   *
+   * @param {string} name name of the mark to record
+   * @param {array} deltaFrom a list of mark names to use as origin
+   * @param {Array} tags array of tag strings
+   * @returns {undefined} undefined
+   */
+  function timing(name, deltaFrom, tags) {
+    if (!deltaFrom) deltaFrom = [];
+    deltaFrom.push('start');
+
+    return function() {
+      // save this time (absolute)
+      times[name] = process.hrtime();
+
+      if (name !== 'start') {
+        // find the first deltaFrom which is set
+        var timeDelta = deltaFrom.reduce(function(result, timeMarkName) {
+          if (!result && times[timeMarkName]) {
+            return timeMarkName;
+          }
+          return result;
+        }, null);
+
+        var delta = process.hrtime(times[timeDelta]);
+        var ms = Math.round(delta[0] * 1000 + delta[1] / 1000000, 2);
+        record(name, ms, tags)
```

Comment:
A really nice pattern that I've settled into with this type of timing data is to record `start`, `end` and `duration`. Without running this code, I'm not sure if you can get that. A good example of this is that I generally like to monitor time to first byte, which is essentially the `firstByte` start timestamp minus the initial timestamp. This metric is represents DNS + TCP + TLS + RESPONSE START. I find it easier to monitor that metric, then dive into the specific composite timings to see what is causing slowness.

I also like the approach of generating these different timings at this level because different engineers will have a different uses for the metrics so it's helpful to just make them available. That said, this work will cause a proliferation of custom metrics, which can be costly.